### DISCUSSION TITLE
Avoid rules() running twice in mount() and thrice on validation.

### DISCUSSION THREAD

**Discussion by tanthammar** [livewire/livewire#4045] [3. Feature ideas]
When a component is mounted the `rules()` method runs 2 times, and another 3 times on `validate()`.

I have some dynamic conditions and a bit heavy calculations to set some of my `$rules`. 
I would like to evaluate this code max, once per request.

Would be great if the `rules()` method could be treated as a computed property to avoid unnecessary repetitions.

Example

web.php
```php
Route::get('/rules', \App\Http\Livewire\Rules::class)
```

rules.blade.php
```blade
<div>
    <button wire:click="btnClick">Validate</button>
</div>
```

Rules.php
```php
<?php

namespace App\Http\Livewire;

use App\User;
use Livewire\Component;

class Rules extends Component
{
    public User $user;

    public function mount()
    {
        $this->user = new User;
    }

    protected function rules()
    {
        ray('why does this run twice on mount, and thrice on validate?');
        return [
            'user.name' => 'required'
        ];
    }


    public function btnClick() {
        $this->validate();
    }


    public function render()
    {
        return view('livewire.rules');
    }
}
```

Temp solution to get around it, with my use case,
```php

// trait ValidatesFields

protected int $rulesIteration = 0;
protected array $cachedRules = [];

public function hydrateValidatesFields()
{
    $this->rulesIteration = 1;
}

protected function rules(): array
{
    return $this->fieldRules();
}

protected function fieldRules(): array
{
    if ($this->rulesIteration < 2) {
        $this->rulesIteration += 1;
        $this->cachedRules = $this->recursiveFieldRules();
    }
    return $this->cachedRules;
}

protected function recursiveFieldRules(...): array
{
    return //recursive loop that returns an array with rules
}
```

**Comment by joshhanley** [comment]
@tanthammar yeah that has bitten me before too. Essentially the reason it is called multiple times, is because the rules are used for hydrating/ dehydrating eloquent models/ eloquent collections, as well as for validation.

Ideally it would be computed, but I don't know if that will cause more issues. My concern is that the rules would be memoized when the rules get called during hydration, and so if properties were updated after that, and you need one of those updated properties to construct your rules array, then the memoized rules would actually be wrong.

But you could work around this in your app at the moment, by memoizing it yourself.

```php
public $rules = null;

public function rules () {
    if (is_null($this->rules) {
        $this->rules = []; // Perform calculation of rules here
    }

    return $this->rules;
}
```

This works as under the hood `getRules` first checks for the method `rules` and calls it if it exists, and if not, then it checks for the property `$rules`.

https://github.com/livewire/livewire/blob/991e5bd8a48e450d23cd55336964f916e92a6464/src/ComponentConcerns/ValidatesInput.php#L79-L85

Hope this helps!

**Reply by tanthammar** [reply]
@joshhanley Thank you for replying. 

Yes, smarter without a counter :)

```php
trait ValidatesFields
{
    protected array $memoizedRules = [];

    //sometimes I want to override this trait, then it's easy to return array_merge($this->fieldRules(), [custom rules])
    protected function rules(): array
    {
        return $this->fieldRules();
    }

    protected function fieldRules(): array
    {
        if ($this->memoizedRules === []) {
            $this->memoizedRules = $this->recursiveFieldRules();
        }
        return $this->memoizedRules;
    }


    protected function recursiveFieldRules(): array
    {
        //return calculated $rules;
    }
}
```

**Reply by joshhanley** [reply]
@tanthammar no worries! 🙂 yeah that looks good!
